### PR TITLE
Add iPhone/iOS USB thethering support

### PR DESCRIPTION
**Pull request by jan-kiszka**
This allows to configure USB-attached iPhone as WAN interface. Currently only pre-integrated into the wrt32x, and the build also still complains about "your configuration is out of sync. Please run make menuconfig, oldconfig or defconfig!". Please review and give hints how to get this into shape!

### CODE REVIEW

**Comment by jan-kiszka** [comment]
Addresses #940.

**Comment by lantis1008** [comment]
I probably won't be able to review this until after next weekend sorry. It is on my todo list though.

**Comment by jan-kiszka** [comment]
I've dropped unneeded usbutils and fixed the UI initialization when the "iph" mode is selected.

**Comment by jan-kiszka** [comment]
Yet another update, adding a kernel patch that apparently only OpenWRT carries to fix connection drops caused by URB overflows. Already poked on LKML why the integration is not making progress.

With this integrated, I've spent now the whole day without glitches on the 5G link my iPhone is proving (until the land-line it fixed...).

File: package/plugin-gargoyle-usb-storage-noshare/Makefile [hunk 1]
```diff
@@ -35,6 +35,7 @@ define Package/plugin-gargoyle-usb-storage-noshare
 	DEPENDS+=+kmod-nls-base +kmod-nls-cp1250 +kmod-nls-cp1251 +kmod-nls-cp437 +kmod-nls-cp775 +kmod-nls-cp850 
 	DEPENDS+=+kmod-nls-cp852 +kmod-nls-cp866 +kmod-nls-iso8859-1 +kmod-nls-iso8859-13 +kmod-nls-iso8859-15 
 	DEPENDS+=+kmod-nls-iso8859-2 +kmod-nls-koi8r +kmod-nls-utf8 +disktype +kmod-fuse +ntfs-3g
+	DEPENDS+=+kmod-usb-net-ipheth +usbmuxd
```
Comment (lantis1008) [inline package/plugin-gargoyle-usb-storage-noshare/Makefile:38]: We shouldn't tie this to usb-storage-noshare

**Review by lantis1008 — CHANGES_REQUESTED** [review]
All over, i'm reasonably happy, with minor changes:
- Don't tie it to usb-storage-noshare
- defconfig the target configs (i'm happy to do this afterwards as a separate commit when i add support to all devices)
- Create a meta package to allow the function to be installed (2.5MB is too much to include by default, even on the biggest targets in my opinion)
- Modularise everything. the ipheth kmod is ok to include into the kernel

See below patch with my suggested changes.
```diff
diff --git a/package/plugin-gargoyle-ipheth-tether/Makefile b/package/plugin-gargoyle-ipheth-tether/Makefile
new file mode 100644
index 00000000..2f7203f0
--- /dev/null
+++ b/package/plugin-gargoyle-ipheth-tether/Makefile
@@ -0,0 +1,48 @@
+#
+# Copyright (C) 2006 OpenWrt.org
+#
+# This is free software, licensed under the GNU General Public License v2.
+# See /LICENSE for more information.
+#
+
+include $(TOPDIR)/rules.mk
+
+PKG_NAME:=plugin_gargoyle_ipheth_tether
+PKG_VERSION:=$(GARGOYLE_VERSION)
+ifeq ($(GARGOYLE_VERSION),)
+     PKG_VERSION:=1.0.0
+endif
+PKG_RELEASE:=1
+
+PKG_BUILD_DIR:=$(BUILD_DIR)/$(PKG_NAME)-$(PKG_VERSION)
+
+include $(INCLUDE_DIR)/package.mk
+
+define Package/plugin-gargoyle-ipheth-tether
+	SECTION:=admin
+	CATEGORY:=Administration
+	SUBMENU:=Gargoyle Web Interface
+	TITLE:=iPhone (and other iOS devices) USB tethering support
+	DEPENDS:=+plugin-gargoyle-usb-storage-noshare +kmod-usb-net-ipheth +usbmuxd
+	MAINTAINER:=Eric Bishop
+	PKGARCH:=all
+endef
+
+define Package/plugin-gargoyle-ipheth-tether/description
+	iPhone (and other iOS devices) USB tethering support
+endef
+
+define Build/Prepare
+endef
+
+define Build/Configure
+endef
+
+define Build/Compile
+endef
+
+define Package/plugin-gargoyle-ipheth-tether/install
+	$(INSTALL_DIR) $(1)
+endef
+
+$(eval $(call BuildPackage,plugin-gargoyle-ipheth-tether))
diff --git a/package/plugin-gargoyle-usb-storage-noshare/Makefile b/package/plugin-gargoyle-usb-storage-noshare/Makefile
index be23151e..93f1fa0a 100644
--- a/package/plugin-gargoyle-usb-storage-noshare/Makefile
+++ b/package/plugin-gargoyle-usb-storage-noshare/Makefile
@@ -35,7 +35,6 @@ define Package/plugin-gargoyle-usb-storage-noshare
 	DEPENDS+=+kmod-nls-base +kmod-nls-cp1250 +kmod-nls-cp1251 +kmod-nls-cp437 +kmod-nls-cp775 +kmod-nls-cp850 
 	DEPENDS+=+kmod-nls-cp852 +kmod-nls-cp866 +kmod-nls-iso8859-1 +kmod-nls-iso8859-13 +kmod-nls-iso8859-15 
 	DEPENDS+=+kmod-nls-iso8859-2 +kmod-nls-koi8r +kmod-nls-utf8 +disktype +kmod-fuse +ntfs-3g
-	DEPENDS+=+kmod-usb-net-ipheth +usbmuxd
 	MAINTAINER:=Eric Bishop
 	PKGARCH:=all
 endef
diff --git a/targets/mvebu/profiles/default/config b/targets/mvebu/profiles/default/config
index 1f2c05e6..7f8f641e 100644
--- a/targets/mvebu/profiles/default/config
+++ b/targets/mvebu/profiles/default/config
@@ -2360,6 +2360,7 @@ CONFIG_PACKAGE_plugin-gargoyle-i18n-SimplifiedChinese-ZH-CN=m
 CONFIG_PACKAGE_plugin-gargoyle-i18n-Slovak-SK=m
 CONFIG_PACKAGE_plugin-gargoyle-i18n-Spanish-ES=m
 CONFIG_PACKAGE_plugin-gargoyle-initd=m
+CONFIG_PACKAGE_plugin-gargoyle-ipheth-tether=m
 CONFIG_PACKAGE_plugin-gargoyle-logread=m
 CONFIG_PACKAGE_plugin-gargoyle-minidlna=y
 CONFIG_PACKAGE_plugin-gargoyle-openvpn=y
@@ -3330,6 +3331,14 @@ CONFIG_SQLITE3_FTS5=y
 CONFIG_SQLITE3_JSON1=y
 CONFIG_SQLITE3_RTREE=y
 # CONFIG_SQLITE3_SESSION is not set
+
+#
+# libimobiledevice
+#
+CONFIG_PACKAGE_libimobiledevice=m
+CONFIG_PACKAGE_libplist=m
+# CONFIG_PACKAGE_libplistcxx is not set
+CONFIG_PACKAGE_libusbmuxd=m
 # CONFIG_PACKAGE_alsa-lib is not set
 # CONFIG_PACKAGE_argp-standalone is not set
 # CONFIG_PACKAGE_libaio is not set
@@ -3471,6 +3480,7 @@ CONFIG_PACKAGE_libuuid=y
 CONFIG_PACKAGE_libvorbis=y
 CONFIG_PACKAGE_libwrap=y
 # CONFIG_PACKAGE_libx264 is not set
+CONFIG_PACKAGE_libxml2=m
 # CONFIG_PACKAGE_linux-atm is not set
 # CONFIG_PACKAGE_p11-kit is not set
 CONFIG_PACKAGE_terminfo=y
@@ -3951,6 +3961,14 @@ CONFIG_PACKAGE_swap-utils=y
 # database
 #
 # CONFIG_PACKAGE_sqlite3-cli is not set
+
+#
+# libimobiledevice
+#
+# CONFIG_PACKAGE_libimobiledevice-utils is not set
+# CONFIG_PACKAGE_libusbmuxd-utils is not set
+# CONFIG_PACKAGE_plistutil is not set
+CONFIG_PACKAGE_usbmuxd=m
 # CONFIG_PACKAGE_adb is not set
 # CONFIG_PACKAGE_bsdiff is not set
 # CONFIG_PACKAGE_bspatch is not set
@@ -4007,8 +4025,3 @@ CONFIG_PACKAGE_usb-modeswitch=y
 # CONFIG_PACKAGE_uuidgen is not set
 # CONFIG_PACKAGE_v4l-utils is not set
 # CONFIG_PACKAGE_whereis is not set
-CONFIG_PACKAGE_usbmuxd=y
-CONFIG_PACKAGE_libimobiledevice=y
-CONFIG_PACKAGE_libplist=y
-CONFIG_PACKAGE_libusbmuxd=y
-CONFIG_PACKAGE_libxml2=y
```

File: package/plugin-gargoyle-usb-storage-noshare/Makefile (hunk 1 above)
Comment (jan-kiszka) [reply package/plugin-gargoyle-usb-storage-noshare/Makefile:38]: Yeah, will try to create a new plugin. My thought was that the other modems are already located here, but I see the point when usbmuxd actually pulls that much (didn't measure yet).

**Comment by jan-kiszka** [comment]
BTW, the last kernel patch is on its way to 4.19-stable now. So it will eventually (on some OpenWRT update) have to be dropped again.

File: package/plugin-gargoyle-usb-storage-noshare/Makefile (hunk 1 above)
Comment (lantis1008) [reply package/plugin-gargoyle-usb-storage-noshare/Makefile:38]: My suggested patch has this already done for you :)

**Comment by lantis1008** [comment]
Hi @jan-kiszka , any thoughts on progressing this one?
At your leisure, just checking in. :)

**Comment by jan-kiszka** [comment]
Had no time yet (and my cable link got fixed, "unfortunately"). Will try to schedule this for next weekend.

**Comment by jan-kiszka** [comment]
Currently finalizing the patches. However, I only measure about 600K additional packages on mvebu, not 2.5M.

**Comment by lantis1008** [comment]
I'll aim to merge this on the weekend.
For my own notes, in the mvebu case:

| Package | Size |
| ------------------------------------------------------------------------- | --------- |
| kmod-usb-net-ipheth_4.14.267-1_arm_cortex-a9_vfpv3-d16.ipk       | 5.1 KB |
| libusbmuxd_2019-03-23-873252dc-2_arm_cortex-a9_vfpv3.ipk        | 14.2 KB |
| usbmuxd_2019-03-04-b1b0bf39-1_arm_cortex-a9_vfpv3.ipk            | 28.3 KB |
| libimobiledevice_2019-02-16-0584aa90-2_arm_cortex-a9_vfpv3.ipk | 44.5 KB |
| libplist_2.0.0-4_arm_cortex-a9_vfpv3.ipk                                             | 21.7 KB |
| libxml2_2.9.9-2_arm_cortex-a9_vfpv3.ipk                                            | 459.7 KB |

**Comment by jan-kiszka** [comment]
Here is the diff from my device after installing those packages:
```
Root Drive 90.828 MBytes Total, 85.840 MBytes Free
```
vs.
```
Root Drive 91.277 MBytes Total, 86.289 MBytes Free
```
Slightly less than the sum of the ipks.

**Comment by lantis1008** [comment]
Thanks again for your efforts on this. I've merged it and enabled it on what i believe are the appropriate profiles. 0.5MB is still a bit steep, but we'll have to see what devices start to fail to build.

**Pull request merged**